static/app/views/issueList/utils/issueViews.tsx

```diff
@@ -0,0 +1,556 @@
+    return {
+      ...queryParams,
+      project: project ?? pageFilters.selection.projects,
+      environment: environment ?? pageFilters.selection.environments,
+      ...normalizeDateTimeParams(pageFilters.selection.datetime),
+    };
+  }, [
+    environment,
+    pageFilters.selection.datetime,
+    pageFilters.selection.environments,
+    pageFilters.selection.projects,
+    project,
+    queryParams,
+  ]);
+
+  // This function is fired upon receiving new views from the backend - it replaces any previously
+  // generated temporary view ids with the permanent view ids from the backend
+  const replaceWithPersistantViewIds = (views: GroupSearchView[]) => {
+    const newlyCreatedViews = views.filter(
+      view => !state.views.find(tab => tab.id === view.id)
+    );
+    if (newlyCreatedViews.length > 0) {
+      dispatch({type: 'UPDATE_VIEW_IDS', newViews: newlyCreatedViews});
+      const currentView = state.views.find(tab => tab.id === viewId);
+
+      if (viewId?.startsWith('_') && currentView) {
+        const matchingView = newlyCreatedViews.find(
+          view =>
+            view.id &&
+            currentView.query === view.query &&
+            currentView.querySort === view.querySort
+        );
+        if (matchingView?.id) {
+          navigate(
+            normalizeUrl({
+              ...location,
+              query: {
+                ...queryParamsWithPageFilters,
+                viewId: matchingView.id,
+              },
+            }),
+            {replace: true}
+          );
+        }
+      }
+    }
+    return;
+  };
+
+  const {mutate: updateViews} = useUpdateGroupSearchViews({
+    onSuccess: replaceWithPersistantViewIds,
+  });
+
+  const debounceUpdateViews = useMemo(
+    () =>
+      debounce((newTabs: IssueView[]) => {
+        if (newTabs) {
+          updateViews({
+            orgSlug: organization.slug,
+            groupSearchViews: newTabs
+              .filter(tab => tab.isCommitted)
+              .map(tab => ({
+                // Do not send over an ID if it's a temporary or default tab so that
+                // the backend will save these and generate permanent Ids for them
+                ...(tab.id[0] !== '_' && !tab.id.startsWith('default')
+                  ? {id: tab.id}
+                  : {}),
+                name: tab.label,
+                query: tab.query,
+                querySort: tab.querySort,
+              })),
+          });
+        }
+      }, 500),
+    [organization.slug, updateViews]
+  );
+
+  const reducer: Reducer<IssueViewsState, IssueViewsActions> = useCallback(
+    (state, action): IssueViewsState => {
+      if (!tabListState) {
+        return state;
+      }
+      // eslint-disable-next-line no-console
+      console.log('executing dispatch', state, action);
+      let newState;
+      switch (action.type) {
+        case 'REORDER_TABS':
+          newState = reorderTabs(state, action);
+          debounceUpdateViews(newState.views);
```

Comment:
Reducers should be pure and avoid side effects. If we need to update views after an action, we should do that in a useEffect

Reply:
If you need to ignore certain updates, you could:

1. If it's possible to discern from the shape of the data, you could look at that in the useEffect and choose not to call debounceUpdateViews in those cases

or 

2. You could wrap the dispatch and the debounce in a single function and call that instead of `dispatch` directly. That would require some organizational changes - you'd probably want to put an `actions` object in the context that has all the action functions